draft_python/mtg_draft_ai/brains.py

```diff
@@ -1,20 +1,163 @@
+    """Attempts to maximize number of synergy edges, for the color pair with the most edges in its pool."""
+
+    def __init__(self, default_ratings, common_neighbors):
+        """Basic init method which does no manipulation of its parameters.
+
+        For most cases, use GreedySynergyPicker.factory(card_list).create() instead.
+
+        Args:
+            default_ratings ({Card: float}): A default rating for each card.
+            common_neighbors ({Card: {Card: List[Card]}}): Generated by all_common_neighbors function
+        """
+        self.default_ratings = default_ratings
+        self.common_neighbors = common_neighbors
+
+    @staticmethod
+    def factory(card_list):
+        """Creates a factory for GreedySynergyPicker instances.
+
+        Performs potentially expensive precomputations, which can be reused to create multiple instances.
+
+        Args:
+            card_list (List[Card]): List of all cards in the cube.
+
+        Returns:
+            Factory: a factory for GreedySynergyPicker instances, with a no-arguments create() method.
+        """
+        syn_graph = synergy.create_graph(card_list)
+
+        # Default rating is currently based on centrality, but likely we could find a better measure.
+        defaults = nx.eigenvector_centrality(syn_graph)
+        for card in card_list:
+            defaults.setdefault(card, 0)
+
+        common_neighbors = all_common_neighbors(syn_graph, card_list)
+
+        kwargs = {'default_ratings': defaults,
+                  'common_neighbors': common_neighbors}
+        return Factory(GreedySynergyPicker, kwargs)
+
+    def pick(self, pack, cards_owned):
+        ranked_candidates = self._ratings(pack, cards_owned)
+
+        # replace full card object with just card name for more readable output
+        printable_candidates = [str(_GSPRating(tup[0].name, *tup[1:])) for tup in ranked_candidates]
+        print('\nrankings: {}'.format('\n'.join(printable_candidates)))
+
+        return pack[0] if len(ranked_candidates) == 0 else ranked_candidates[0][0]
+
+    def _ratings(self, pack, cards_owned):
+        candidates = []
+
+        for color_combo in COLOR_PAIRS:
+            on_color_cards = [c for c in cards_owned if synergy.castable(c, color_combo)]
+            on_color_candidates = [c for c in pack if synergy.castable(c, color_combo)]
+
+            for candidate in on_color_candidates:
+                cards_with_candidate = on_color_cards + [candidate]
+                syn_graph = synergy.create_graph(cards_with_candidate)
+
+                # Number of edges in the graph for the card pool plus the candidate.
+                total_edges = len(syn_graph.edges)
+                # Number of edges added by the candidate.
+                new_edges = _num_new_edges(syn_graph, candidate)
+                # Number of nodes in global graph which neighbor both the candidate and the card pool.
+                # Does not count cards already in the pool as neighbors.
```

Comment:
I'm not sure why we would choose not to count the cards we already own as neighbors; isn't our intent to find the best match for the cards we already own? Wouldn't that imply we should be using them?

Reply:
I was thinking that they are already counted in the first measure (total # of edges), so this would essentially be double counting them. Like I mentioned in the chat, I think this metric needs to be changed significantly anyway.